lib/cabal-project-parser.nix

```diff
@@ -95,10 +95,90 @@ let
           otherText = "\nsource-repository-package\n" + block;
         }
         else {
-          sourceRepo = extractRepoData cabalProjectFileName lookupSha256 attrs;
+          sourceRepo = extractSourceRepoPackageData cabalProjectFileName lookupSha256 attrs;
           otherText = pkgs.lib.strings.concatStringsSep "\n" x.snd;
         };
 
+  parseSourceRepositoryPackages = cabalProjectFileName: lookupSha256: source-repo-override: projectFile:
+    let
+      blocks = pkgs.lib.splitString "\nsource-repository-package\n" ("\n" + projectFile);
+      initialText = pkgs.lib.lists.take 1 blocks;
+      repoBlocks = builtins.map (parseBlock cabalProjectFileName lookupSha256) (pkgs.lib.lists.drop 1 blocks);
+      overrideSourceRepo = sourceRepo: (source-repo-override.${sourceRepo.url} or (pkgs.lib.id)) sourceRepo;
+    in {
+      sourceRepos = pkgs.lib.lists.map (block: overrideSourceRepo block.sourceRepo) repoBlocks;
+      otherText = pkgs.lib.strings.concatStringsSep "\n" (
+        initialText
+        ++ (builtins.map (x: x.otherText) repoBlocks));
+    };
+
+  # Parse a repository
+  parseRepositoryBlock = cabalProjectFileName: sha256map: cabal-install: nix-tools: block:
+    let
+      lines = pkgs.lib.splitString "\n" block;
+      x = span (pkgs.lib.strings.hasPrefix " ") (__tail lines);
+      attrs = parseBlockLines x.fst;
+    in rec {
+      name = __head lines;
+      repo = attrs;
+      home =
+        pkgs.evalPackages.runCommandLocal name {
```

Comment:
It's a bit inconsistent that we build the derivation here, whereas for source-repository-packages we just return the data about the repo and fetch it later. Could we put them both in the same place?

Reply:
I did not like that either.  I think the `source-repository-package` stuff needs to be tidied up a bit and I took another look.  It gets tricky because the `/nix/store` paths from the replaced `source-repository-packages` leak into the `plan-nix` (the repository blocks do not have this problem as only the package IDs wind up in the plan.json).  I think we might be able to improve the way we deal with `source-repository-package`, but I think it would be better in another PR.